src/cephadm/cephadm

```diff
@@ -3479,130 +3615,160 @@ def list_daemons(detail=True, legacy_dir=None):
-                                        version = err.split(' ')[2]
-                                        seen_versions[image_id] = version
-                                else:
-                                    logging.warning('version for unknown daemon type %s' % daemon_type)
+                    if daemon_name in daemon_lookup:
+                        _container = daemon_lookup[daemon_name]
+                        daemon_type = daemon_name.split('.')[0]
+
+                        # The goal is to provide version information for each daemon - but this is expensive,
+                        # when the only way of doing it is with an 'exec' command. Instead, we get the version
+                        # info from API endpoints for monitoring, and leave the main ceph versions empty.
+                        # The orch/cephadm module can then fill in the ceph_version information from in-memory
+                        # metadata
+                        method = version_call.get(daemon_type, None)  
+                        if method:
+                            if daemon_type in [CephIscsi.daemon_type, NFSGanesha.daemon_type]:
+                                version = method(_container['ID'])  # type: ignore
+                            elif daemon_type == 'alertmanager':
+                                version = method(_container['Image'])  # type: ignore
+                            else:
+                                # prometheus, grafana, node-exporter
+                                version = method()  # type: ignore
                         else:
-                            vfile = os.path.join(data_dir, fsid, j, 'unit.image') # type: ignore
-                            try:
-                                with open(vfile, 'r') as f:
-                                    image_name = f.read().strip() or None
-                            except IOError:
-                                pass
-                        i['container_id'] = container_id
-                        i['container_image_name'] = image_name
-                        i['container_image_id'] = image_id
-                        i['version'] = version
-                        i['started'] = start_stamp
-                        i['created'] = get_file_timestamp(
-                            os.path.join(data_dir, fsid, j, 'unit.created')
-                        )
-                        i['deployed'] = get_file_timestamp(
-                            os.path.join(data_dir, fsid, j, 'unit.image'))
-                        i['configured'] = get_file_timestamp(
-                            os.path.join(data_dir, fsid, j, 'unit.configured'))
-
-                    ls.append(i)
+                            if daemon_type == 'crash':
+                                # yield octopus, master
+                                version = _container['Env'].get('CEPH_VERSION', 'Unknown')
+                            elif daemon_type in Ceph.daemons:
```

Comment:
To share one idea:
Maybe we should add a label  with the version in all the daemons deployed using cephadm.deploy_daemon. When we run  "deploy_daemon " we know what is the image we are using , and probably is the best  moment to find out the version of the daemon and write this information as a label in the container (also we capn include the version as mandatory parameter in the deploy daemon function) . This probably can simplify a lot the code for getting versions. 
In any case , is something that we can try in other PR.

Reply:
Yep - labels is where  was thinking too - but the labels are set at container build time, and the current build process does not allow the labels to include the ceph version from the components being built into the container.